pipeline-model-definition/src/main/java/org/jenkinsci/plugins/pipeline/modeldefinition/actions/RestartDeclarativePipelineAction.java

```diff
@@ -0,0 +1,207 @@
+import org.jenkinsci.plugins.workflow.flow.FlowExecution;
+import org.jenkinsci.plugins.workflow.flow.FlowExecutionOwner;
+import org.jenkinsci.plugins.workflow.job.WorkflowRun;
+import org.kohsuke.accmod.Restricted;
+import org.kohsuke.accmod.restrictions.DoNotUse;
+import org.kohsuke.stapler.QueryParameter;
+import org.kohsuke.stapler.StaplerRequest;
+import org.kohsuke.stapler.StaplerResponse;
+import org.kohsuke.stapler.interceptor.RequirePOST;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import javax.servlet.ServletException;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.List;
+
+import static javax.servlet.http.HttpServletResponse.SC_BAD_REQUEST;
+import static javax.servlet.http.HttpServletResponse.SC_CONFLICT;
+
+public class RestartDeclarativePipelineAction implements Action {
+
+    private final Run run;
+
+    private RestartDeclarativePipelineAction(Run run) {
+        this.run = run;
+    }
+
+    @Override public String getDisplayName() {
+        return "Restart from Stage";
+    }
+
+    @Override public String getIconFileName() {
+        // TODO: Needs an icon.
+        return null;
+    }
+
+    @Override public String getUrlName() {
+        return isRestartEnabled() ? "restart" : null;
+    }
+
+    /* accessible to Jelly */ public Run getOwner() {
+        return run;
+    }
+
+    @CheckForNull
+    private CpsFlowExecution getExecution() {
+        FlowExecutionOwner owner = ((FlowExecutionOwner.Executable) run).asFlowExecutionOwner();
+        if (owner == null) {
+            return null;
+        }
+        FlowExecution exec = owner.getOrNull();
+        return exec instanceof CpsFlowExecution ? (CpsFlowExecution) exec : null;
+    }
+
+    public boolean isRestartEnabled() {
+        ExecutionModelAction executionModelAction = run.getAction(ExecutionModelAction.class);
+
+        return executionModelAction != null &&
+                !run.isBuilding() &&
+                run.hasPermission(Item.BUILD) &&
+                run.getParent().isBuildable() &&
+                getExecution() != null;
+    }
+
+    @Restricted(DoNotUse.class)
+    @RequirePOST
+    public void doRestart(StaplerRequest req, StaplerResponse rsp) throws ServletException, IOException {
+        if (!isRestartEnabled()) {
+            throw new AccessDeniedException("not allowed to restart"); // AccessDeniedException2 requires us to look up the specific Permission
+        }
+        JSONObject form = req.getSubmittedForm();
+        String stageName = Util.fixEmpty(form.getString("stageName"));
+
+        try {
+            run(stageName);
+        } catch (IllegalStateException ise) {
+            throw HttpResponses.error(SC_CONFLICT, ise);
+        }
+
+        rsp.sendRedirect("../.."); // back to WorkflowJob; new build might not start instantly so cannot redirect to it
```

Comment:
Make a mention of this case in [JENKINS-49138](https://issues.jenkins-ci.org/browse/JENKINS-49138).